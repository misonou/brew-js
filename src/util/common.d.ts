export function getAttrValues(element: Element): Zeta.Dictionary<string>;

/**
 * Gets whether a given attribute is a boolean-mapped attribute for an element.
 * @param element A DOM element.
 * @param name Name of attribute to test.
 */
export function isBoolAttr(element: Element, name: string): boolean;

export function hasAttr(element: Element, name: string): boolean;

export function getAttr(element: Element, name: string): string | undefined;

export function setAttr(element: Element, name: string, value: string | null): void;

export function setAttr(element: Element, values: Zeta.Dictionary<string | null>): void;

export function copyAttr(src: Element, dst: Element): void;

export function selectorForAttr(attr: string[] | Zeta.Dictionary): string;

export function getFormValues(form: HTMLFormElement): Zeta.Dictionary;

export function getQueryParam(name: string, current?: string): string | false;

export function setQueryParam(name: string, value: string, current?: string): string;

export function getCookie(name: string): string;

export function setCookie(name: string, value: string, expiry?: number): string;

export function deleteCookie(name: string): void;

export function cookie(name: string, expiry?: number): {
    get: () => string;
    set: (value: any) => string;
    delete: () => void;
};

export function api(options?: Brew.APIOptions): Brew.API;

export function api(method: Brew.HTTPMethod, baseUrl: string): Brew.APIMethod;

export function api(method: Brew.HTTPMethod, options?: Omit<Brew.APIOptions<T>, 'methods'>): Brew.APIMethod;

export function getJSON(path: string): Promise<Zeta.Dictionary>;

export function loadScript(url: string | string[], options?: {
    nomodule?: boolean;
    module?: boolean;
}): Promise<Zeta.Dictionary>;

export function addStyleSheet(url: string, media?: string): void;

export function preloadImages(urls: string[] | Element, ms?: number): Promise<any>;
