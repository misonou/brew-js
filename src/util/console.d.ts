export function writeLog(eventSource: string, message: string | Element | Record<any, any> | any[]): string;

export function groupLog(eventSource: string, message: string | Element | Record<any, any> | any[], callback: (console: Console) => void): void;
