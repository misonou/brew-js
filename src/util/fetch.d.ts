import * as ErrorCode from "../errorCode.js";

export type FetchLayer = (req: Request) => Promise<Response>;
export type FetchMiddleware = (req: Request, next: FetchLayer) => Promise<Response>;
export type RequestOptions = Omit<RequestInit, 'method' | 'body'>;

export interface ApiClientError extends Error {
    /**
     * Gets the type of error.
     *
     * Possible values include
     * - {@link ErrorCode.networkError}
     * - {@link ErrorCode.apiError}
     */
    readonly code: string;
    /**
     * Gets the HTTP status code responded from server.
     */
    readonly status: number;
    /**
     * Gets data parsed from response body; or `undefined` when body cannot be parsed.
     */
    readonly data: any;
    /**
     * Gets error thrown when parsing response body; or `null` when body is parsed successfully.
     */
    readonly error: any;
}

export interface ApiClientTraps {
    /**
     * Specifies callback to transform request or response.
     *
     * The callback is invoked before fetch middlewares registered by {@link ApiClient.use},
     * and the promise returned from `next` callback
     */
    request?: (req: Request, next: (req: Request) => Promise<any>) => any;
    /**
     * Specifies callback to receives and transform data.
     * @param data Data deserialized from response.
     * @param response Response object.
     * @param request Request object associated with the response.
     * @param client Client object that invoked the request.
     */
    data?: (data: any, response: Response, request: Request, client: ApiClient) => any;
    /**
     * Specifies callback to handle request error.
     *
     * The trap is invoked when
     * - response status indicates an HTTP error (400-599), or
     * - failed to parse response payload,
     * - failed to execute request, for example, due to network or CORS issue.
     *
     * If non-`undefined` value is returned, request will be treated as successful and
     * promise will be settled with the returned value. Otherwise the incoming error will be rethrown.
     *
     * @param error Error object.
     * @param response Response object.
     * @param request Request object associated with the response.
     * @param client Client object that invoked the request.
     */
    error?: (error: ApiClientError, response: Response, request: Request, client: ApiClient) => any;
}

export interface ApiClient {
    /**
     * Makes request to an endpoint.
     */
    <T = any>(path: RequestInfo | URL, init?: RequestInit): Promise<T>;
    /**
     * Makes request to an endpoint.
     */
    <T = any>(method: Brew.HTTPMethod, path: string, body?: any, options?: RequestOptions): Promise<T>;

    /**
     * Pushes a middleware into fetch function.
     * @param middleware A callback that infiltrate the request to the last registered middleware.
     *
     * @example
     * ```javascript
     * // injecting header
     * client.use((req, next) => {
     *   req.headers.set('authorization', 'xxxxx');
     *   return next(req);
     * });
     *
     * // transforming response
     * client.use((req, next) => {
     *   return next(req).then(res => {
     *     return res.json().then(data => {
     *       return Response.json(data.d, res);
     *     });
     *   });
     * });
     * ```
     */
    use(middleware: FetchMiddleware): void;

    /**
     * Makes a `GET` request to an endpoint.
     */
    get<T = any>(path: string, options?: RequestOptions): Promise<T>;
    /**
     * Makes a `HEAD` request to an endpoint.
     */
    head<T = any>(path: string, options?: RequestOptions): Promise<T>;
    /**
     * Makes a `POST` request to an endpoint.
     */
    post<T = any>(path: string, body?: any, options?: RequestOptions): Promise<T>;
    /**
     * Makes a `PUT` request to an endpoint.
     */
    put<T = any>(path: string, body?: any, options?: RequestOptions): Promise<T>;
    /**
     * Makes a `PATCH` request to an endpoint.
     */
    patch<T = any>(path: string, body?: any, options?: RequestOptions): Promise<T>;
    /**
     * Makes a `DELETE` request to an endpoint.
     */
    delete<T = any>(path: string, body?: any, options?: RequestOptions): Promise<T>;
}

/**
 * Creates a fetch function that is interceptable with middlewares.
 * @param baseUrl Base URL prepend to relative request path.
 * @param traps A dictionary containing intercepting callback.
 * @description
 * Below illustrates the invocation order of traps and middlewares.
 * ```text
 * request()
 *  |  Λ
 *  V  |
 * traps.request
 *  |  Λ
 *  |  |
 *  |  traps.data (or traps.error)
 *  |  Λ
 *  V  |
 * middleware (last)
 *  |  Λ
 *  :  :
 *  V  |
 * middleware (first)
 *  |  Λ
 *  V  |
 * native fetch
 * ```
 *
 * Client will parse response body according to `Accept` header from request, and fallback to `Content-Type` header from response.
 * See {@link peekBody} for supported header values and associated data types.
 */
export function createApiClient(baseUrl: string, traps?: ApiClientTraps): ApiClient;

/**
 * Reads request or response body.
 * @param input Request or response object. Source body stream is untouched.
 * @returns A promise that resolves to parsed data according to the `Content-Type` header:
 * - deserialized JSON data for `application/json`;
 * - deserialized data as a `FormData` object for `multipart/form-data` or `application/x-www-form-urlencoded`;
 * - decoded string for `text/*`;
 * - binary data as a `Blob` object otherwise.
 */
export function peekBody(input: Request | Response): Promise<any>;
