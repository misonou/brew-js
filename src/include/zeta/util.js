export * from "zeta-dom/util.js";
