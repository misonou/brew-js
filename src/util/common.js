import $ from "../include/external/jquery.js";
import Promise from "../include/external/promise-polyfill.js";
import { isCssUrlValue } from "../include/zeta-dom/cssUtil.js";
import { createNodeIterator, iterateNode, matchSelector } from "../include/zeta-dom/domUtil.js";
import { defineAliasProperty, defineGetterProperty, defineHiddenProperty, each, errorWithCode, extend, isArray, isFunction, isPlainObject, keys, kv, matchWord, resolve, resolveAll, setPromiseTimeout, values, watchOnce } from "../include/zeta-dom/util.js";
import { combinePath, withBaseUrl } from "./path.js";
import * as ErrorCode from "../errorCode.js";

/** @type {Zeta.Dictionary<Promise<void>>} */
const preloadImagesCache = {};
/** @type {Zeta.Dictionary<Promise<Zeta.Dictionary>>} */
const loadScriptCache = {};

export function getAttrValues(element) {
    var values = {};
    each(element.attributes, function (i, v) {
        values[v.name] = v.value;
    });
    return values;
}

export function isBoolAttr(element, name) {
    return matchWord(name, 'allowfullscreen async autofocus autoplay checked controls default defer disabled formnovalidate ismap itemscope loop multiple muted nomodule novalidate open playsinline readonly required reversed selected truespeed') && name in element;
}

export function hasAttr(element, name) {
    return !!element.attributes[name];
}

export function getAttr(element, name) {
    var attr = element.attributes[name];
    return attr && attr.value;
}

export function setAttr(element, name, value) {
    each(isPlainObject(name) || kv(name, value), function (i, v) {
        element.setAttribute(i, v);
    });
}

export function copyAttr(src, dst) {
    each(src.attributes, function (i, v) {
        dst.setAttribute(v.name, v.value);
    });
}

export function selectorForAttr(attr) {
    if (isPlainObject(attr)) {
        attr = keys(attr);
    }
    return attr[0] ? '[' + attr.join('],[') + ']' : '';
}

/**
 * @param {HTMLFormElement} form
 */
export function getFormValues(form) {
    var values = {};
    each(form.elements, function (i, v) {
        if (v.name && !(v.name in values)) {
            var item = form.elements[v.name];
            defineGetterProperty(values, v.name, function () {
                return item.type === 'checkbox' ? item.checked : item.value;
            });
        }
    });
    return values;
}

/**
 * @param {string} name
 */
export function getQueryParam(name) {
    return new RegExp('[?&]' + name + '=([^&]+)', 'i').test(location.search) && decodeURIComponent(RegExp.$1);
}

/**
 * @param {string} name
 */
export function getCookie(name) {
    return new RegExp('(?:^|\\s|;)' + name + '=([^;]+)').test(document.cookie) && RegExp.$1;
}

/**
 * @param {string} name
 * @param {string} value
 * @param {number=} expiry
 */
export function setCookie(name, value, expiry) {
    document.cookie = name + '=' + value + ';path=/' + (expiry ? ';expires=' + new Date(Date.now() + expiry).toGMTString() : '');
    return value;
}

/**
 * @param {string} name
 */
export function deleteCookie(name) {
    document.cookie = name + '=;path=/;expires=Thu, 01 Jan 1970 00:00:01 GMT';
}

/**
 * @param {string} name
 * @param {number=} expiry
 */
export function cookie(name, expiry) {
    return {
        get: function () {
            return getCookie(name);
        },
        set: function (value) {
            return setCookie(name, value, expiry);
        },
        delete: function () {
            return deleteCookie(name);
        }
    };
}

/**
 * @param {Brew.APIMethod | Brew.APIOptions} options
 * @param {=} extra
 */
export function api(options, extra) {
    var httpMethods = 'get post delete';
    if (typeof options === 'string' && matchWord(options, httpMethods)) {
        extra = extend({}, typeof extra === 'string' ? { baseUrl: extra } : extra, { methods: options });
        return api(extra)[options];
    }
    options = extend({}, options);
    var obj = {
        baseUrl: options.baseUrl,
        token: options.token
    };
    each(options.methods || httpMethods, function (i, v) {
        if (matchWord(v, httpMethods)) {
            defineHiddenProperty(obj, v, function request(method, data) {
                if (!obj.baseUrl) {
                    return watchOnce(obj, 'baseUrl', function () {
                        return request(method, data);
                    });
                }
                var headers = { 'Content-Type': 'application/json' };
                if (obj.token) {
                    headers.Authorization = 'Bearer ' + obj.token;
                }
                return $.ajax({
                    method: v,
                    url: combinePath(obj.baseUrl, method),
                    headers: headers,
                    dataType: 'json',
                    data: JSON.stringify(data || {}),
                    success: function (response) {
                        if (isFunction(options.getTokenFromResponse)) {
                            obj.token = options.getTokenFromResponse(response, obj.token);
                        }
                    }
                }).catch(function (e) {
                    if (e.status === 0) {
                        throw errorWithCode(ErrorCode.networkError);
                    }
                    var response = e.responseJSON;
                    if (response) {
                        console.error(method + ':', response.error || response.message);
                        throw errorWithCode(ErrorCode.apiError, response.error || response.message);
                    }
                    throw errorWithCode(ErrorCode.apiError, e.statusText);
                });
            });
            defineAliasProperty(obj[v], 'baseUrl', obj);
            defineAliasProperty(obj[v], 'token', obj);
        }
    });
    return obj;
}

/**
 * @param {string} path
 */
export function getJSON(path) {
    return $.getJSON(withBaseUrl(path));
}

/**
 * @param {string | string[]} url
 * @param {{ nomodule?: boolean; module?: boolean }=} options
 */
export function loadScript(url, options) {
    if (isArray(url)) {
        return url.reduce(function (v, a) {
            return v.then(function () {
                return loadScript(a, options);
            });
        }, resolve());
    }
    if (!loadScriptCache[url]) {
        loadScriptCache[url] = new Promise(function (resolve, reject) {
            var script = document.createElement('script');
            options = options || {};
            if (options.nomodule) {
                if ('noModule' in script) {
                    resolve();
                    return;
                }
                script.setAttribute('nomodule', '');
            }
            if (options.module) {
                script.setAttribute('type', 'module');
            }
            script.addEventListener('load', function () {
                resolve({});
            });
            script.addEventListener('error', function () {
                reject(errorWithCode(ErrorCode.resourceError));
            });
            script.src = withBaseUrl(url);
            document.head.appendChild(script);
        });
    }
    return loadScriptCache[url];
}

/**
 * @param {string} url
 * @param {string=} media
 */
export function addStyleSheet(url, media) {
    var link = document.createElement('link');
    link.rel = 'stylesheet';
    link.href = withBaseUrl(url);
    if (media) {
        link.media = media;
    }
    document.head.appendChild(link);
}

/**
 * @param {string[]|Element} urls
 * @param {number=} ms
 */
export function preloadImages(urls, ms) {
    if (!isArray(urls)) {
        var map = {};
        var testValue = function (value) {
            if (isCssUrlValue(value)) {
                map[RegExp.$1 || RegExp.$2 || RegExp.$3] = true;
            }
        };
        iterateNode(createNodeIterator(urls, 1), function (node) {
            if (matchSelector(node, 'img') && node.src) {
                map[node.src] = true;
            }
            testValue(getComputedStyle(node).backgroundImage);
            testValue(getComputedStyle(node, '::before').backgroundImage);
            testValue(getComputedStyle(node, '::after').backgroundImage);
        });
        urls = keys(map);
    }
    var promises = [];
    var preloadUrls = [];
    urls.forEach(function (url) {
        promises.push(preloadImagesCache[url] || (preloadImagesCache[url] = new Promise(function (resolve) {
            preloadUrls.push(url);
            $('<img>').on('load error', function () {
                preloadImagesCache[url] = true;
                resolve();
            }).attr('src', url);
        })));
    });
    if (!promises.length || promises.every(function (v) { return v === true; })) {
        return resolve();
    }
    if (preloadUrls.length) {
        console.log('Preload image', { urls: preloadUrls });
    }
    return setPromiseTimeout(resolveAll(values(preloadImagesCache)), ms, true);
}
