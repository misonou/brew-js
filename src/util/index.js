export * from "./common.js";
export * from "./console.js";
export * from "./fetch.js";
export * from "./path.js";
export * from "./storage.js";
