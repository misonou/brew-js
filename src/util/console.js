import { is, makeArray } from "zeta-dom/util";

function toElementTag(element) {
    return element.tagName.toLowerCase() + (element.id ? '#' + element.id : element.className.trim() && element.className.replace(/^\s*|\s+(?=\S)/g, '.'));
}

function truncateJSON(json) {
    return '[{"]'.indexOf(json[0]) >= 0 && json.length > 200 ? json[0] + json.substr(1, 199) + '\u2026' + json[json.length - 1] : json;
}

function formatMessage(eventSource, message) {
    message = makeArray(message).map(function (v) {
        return is(v, Element) ? toElementTag(v) + ':' : v && typeof v === 'object' ? truncateJSON(JSON.stringify(v)) : v;
    }).join(' ');
    return '[' + eventSource + '] ' + message;
}

/**
 * @param {string} eventSource
 * @param {string | Element | Record<any, any> | any[]} message
 */
export function writeLog(eventSource, message) {
    console.log(formatMessage(eventSource, message));
}

/**
 * @param {string} eventSource
 * @param {string | Element | Record<any, any> | any[]} message
 * @param {(console: Console) => void} callback
 */
export function groupLog(eventSource, message, callback) {
    var close;
    try {
        console.groupCollapsed(formatMessage(eventSource, message));
        close = true;
    } catch (e) { }
    try {
        callback(console);
    } finally {
        if (close) {
            try {
                console.groupEnd();
            } catch (e) { }
        }
    }
}
