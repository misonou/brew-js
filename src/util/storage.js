import { compressToUTF16, decompressFromUTF16 } from "../include/lz-string.js";
import { each, extend, is, keys, map, noop, setImmediateOnce, throws } from "zeta-dom/util";

const UNDEFINED = 'undefined';

function isObject(value) {
    return value && typeof value === 'object';
}

function shouldIgnore(obj) {
    return obj === window || is(obj, RegExp) || is(obj, Blob) || is(obj, Node);
}

export function createObjectStorage(storage, key) {
    var types = new Map([
        [Date, '#Date '],
        ['Date', [Date, noop]]
    ]);
    var objectCache = {};
    var objectMap = new WeakMap();
    var dirty = new Set();
    var entries = Object.create(null);
    var serialized = initFromStorage(entries) || ['{}'];
    var emptyIds = map(serialized, function (v, i) {
        return v ? null : i;
    });

    function initFromStorage(entries) {
        try {
            var serialized = decompressFromUTF16(storage.getItem(key)).split('\n');
            extend(entries, JSON.parse(serialized[0]));
            return serialized;
        } catch (e) { }
    }

    function getNextId() {
        return emptyIds.length ? emptyIds.shift() : serialized.push('') - 1;
    }

    function getNextIdForKey(key) {
        var id = entries[key];
        if (id && !(id in objectCache ? isObject(objectCache[id]) : /^[\[\{]/.test(serialized[id]))) {
            // reuse existing index only if target is not an object as it may still be referenced
            return id;
        }
        return getNextId();
    }

    function cacheObject(id, obj) {
        objectCache[id] = obj;
        if (isObject(obj)) {
            objectMap.set(obj, +id);
        }
    }

    function uncacheObject(id) {
        objectMap.delete(objectCache[id]);
        delete objectCache[id];
    }

    function serialize(obj, visited) {
        var counter = 0;
        var str = JSON.stringify(obj, function (k, v) {
            if (!isObject(v)) {
                return typeof v === 'string' && v[0] === '#' ? '#' + v : v;
            }
            if (shouldIgnore(v)) {
                return counter ? undefined : {};
            }
            if (!counter++) {
                return v;
            }
            var id = objectMap.get(v) || getNextId();
            cacheObject(id, v);
            if (!visited[id]) {
                visited[id] = true;
                serialized[id] = serialize(v, visited);
                dirty.delete(v);
            }
            return '#' + id;
        });
        var prefix = obj && typeof obj === 'object' && types.get(obj.constructor);
        return (prefix || '') + str;
    }

    function deserialize(str, refs) {
        if (!str || str === UNDEFINED) {
            return;
        }
        if (str[0] === '#') {
            var pos = str.indexOf(' ');
            var type = types.get(str.slice(1, pos));
            str = str.slice(pos + 1);
            if (type) {
                var len = refs.length;
                var data = deserialize(str, refs);
                if (refs.length === len) {
                    return new type[0](data);
                } else {
                    var target = new type[0]();
                    refs.push({ t: type[1].bind(undefined, target, data) });
                    return target;
                }
            }
        }
        return JSON.parse(str, function (k, v) {
            if (typeof v === 'string' && v[0] === '#') {
                v = v.slice(1);
                if (v[0] !== '#') {
                    if (!(v in objectCache)) {
                        objectCache[v] = undefined;
                        cacheObject(v, deserialize(serialized[v], refs));
                    } else if (objectCache[v] === undefined) {
                        refs.push({ o: this, k, v });
                    }
                    return objectCache[v];
                }
            }
            return v;
        });
    }

    function revive(key, callback) {
        var id = entries[key];
        if (id && serialized[id] && !(id in objectCache)) {
            try {
                var refs = [];
                var value = deserialize('{"": "#' + id + '"}', refs)[""];
                each(refs, function (i, v) {
                    if (v.t) {
                        v.t();
                    } else {
                        v.o[v.k] = objectCache[v.v];
                    }
                });
                if (callback) {
                    uncacheObject(id);
                    cacheObject(id, callback(value));
                }
            } catch (e) {
                serialized[id] = UNDEFINED;
            }
        }
        return objectCache[id];
    }

    function persist() {
        var visited = {};
        each(dirty, function (i, v) {
            var id = objectMap.get(v);
            if (id) {
                serialized[id] = serialize(v, visited);
            }
        });
        visited = { 0: true };
        each(entries, function visit(_, id) {
            if (!visited[id]) {
                visited[id] = true;
                serialized[id].replace(/[\[:,]"#(\d+)"/g, visit);
            }
        });
        each(serialized, function (i) {
            if (!visited[i] && serialized[i]) {
                serialized[i] = '';
                emptyIds.push(i);
                uncacheObject(i);
            }
        });
        dirty.clear();
        serialized[0] = JSON.stringify(entries);
        storage.setItem(key, compressToUTF16(serialized.join('\n').trim()));
    }

    return {
        registerType: function (key, fn, setter) {
            if (types.has(key) || types.has(fn)) {
                throws('Key or constructor already registered');
            }
            types.set(fn, '#' + key + ' ');
            types.set(key, [fn, setter]);
        },
        keys: function () {
            return keys(entries);
        },
        has: function (key) {
            return !!entries[key];
        },
        get: function (key) {
            return revive(key);
        },
        revive: function (key, callback) {
            uncacheObject(entries[key]);
            return revive(key, callback);
        },
        set: function (key, value) {
            var id = objectMap.get(value) || getNextIdForKey(key);
            entries[key] = id;
            if (!isObject(value)) {
                serialized[id] = serialize(value) || UNDEFINED;
            } else {
                cacheObject(id, value);
                dirty.add(value);
            }
            setImmediateOnce(persist);
        },
        persist: function (obj) {
            dirty.add(obj);
            setImmediateOnce(persist);
        },
        persistAll: function () {
            each(entries, function (i, v) {
                if (isObject(objectCache[v])) {
                    dirty.add(objectCache[v]);
                }
            });
            setImmediateOnce(persist);
        },
        delete: function (key) {
            if (entries[key]) {
                delete entries[key];
                setImmediateOnce(persist);
            }
        },
        clear: function () {
            serialized.splice(0, serialized.length, '{}');
            emptyIds.splice(0);
            dirty.clear();
            objectMap = new WeakMap();
            objectCache = {};
            entries = Object.create(null);
            setImmediateOnce(persist);
        }
    };
}
