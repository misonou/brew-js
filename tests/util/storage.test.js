import { createObjectStorage } from "src/util/storage";
import { delay, mockFn } from "../testUtil";

const TEST_KEY = '__test__';

beforeEach(() => {
    sessionStorage.clear();
});

describe('ObjectStorage', () => {
    it('should clear all data if entry dictionary is corrupted', async () => {
        sessionStorage.setItem(TEST_KEY, '{a}\n{}\n{}');
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        expect(store.keys()).toEqual([]);
    });

    it('should return undefined for corrupted entry', async () => {
        sessionStorage.setItem(TEST_KEY, [
            '{"foo":1,"bar":2}',
            '{a}',
            '{}'
        ].join('\n'));

        const store = createObjectStorage(sessionStorage, TEST_KEY);
        expect(store.get('foo')).toBeUndefined();
    });
});

describe('ObjectStorage.has', () => {
    it('should return boolean indicating presence of specified key', async () => {
        const data = { foo: 'bar' };
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('test', data);
        expect(store.has('test')).toBe(true);
        expect(store.has('foo')).toBe(false);
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        expect(store2.has('test')).toBe(true);
        expect(store2.has('foo')).toBe(false);
    });
});

describe('ObjectStorage.set', () => {
    it('should serialize object to backing storage', async () => {
        const data = { foo: 'bar' };
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('test', data);
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        const restored = store2.get('test');
        expect(restored).toEqual(data);
        expect(restored).not.toBe(data);
        expect(sessionStorage.key(0)).toBe(TEST_KEY);
    });

    it('should serialize object with circular reference', async () => {
        const a = {};
        const b = {};
        a.b = b;
        b.a = a;

        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('test', { a });
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        const data = store2.get('test');
        expect(data.a).toBe(data.a.b.a);
    });

    it('should serialize primitive value except symbol', async () => {
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('string', '#foo\nbar');
        store.set('number', 1);
        store.set('boolean', true);
        store.set('null', null);
        store.set('undefined', undefined);
        expect(store.keys().length).toBe(5);
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        expect(store2.get('string')).toBe('#foo\nbar');
        expect(store2.get('number')).toBe(1);
        expect(store2.get('boolean')).toBe(true);
        expect(store2.get('null')).toBe(null);
        expect(store2.get('undefined')).toBe(undefined);
        expect(store.keys().length).toBe(5);
    });

    it('should not serialize certain non-trival objects', async () => {
        // require in jsdom as stringify(document.body) always gives '{}'
        document.body[TEST_KEY] = 1;

        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('data', {
            blob: new Blob([new Uint8Array(100)]),
            node: document.body,
            window: window
        });
        store.set('blob', new Blob());
        store.set('node', document.body);
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        expect(store2.get('data')).toEqual({});
        expect(store2.get('blob')).toEqual({});
        expect(store2.get('node')).toEqual({});
    });

    it('should call toJSON on object', async () => {
        const toJSON = mockFn().mockReturnValue({ a: 1 });
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('test', { toJSON });
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        expect(store2.get('test')).toEqual({ a: 1 });
        expect(toJSON).toBeCalledTimes(1);
    });
});

describe('ObjectStorage.revive', () => {
    it('should invoke callback and cache the returned value', async () => {
        const data = { foo: 'bar' };
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('test', data);
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        const cb = mockFn(v => ({ ...v }));
        const restored = store2.revive('test', cb);
        expect(cb).toBeCalledTimes(1);
        expect(restored).toBe(cb.mock.results[0].value);
        expect(restored).toEqual(data);
        expect(store2.get('test')).toBe(restored);
    });

    it('should persist object returned by the callback', async () => {
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('test', { foo: 'bar' });
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        const cb = mockFn(v => ({ ...v }));
        const restored = store2.revive('test', cb);

        restored.foo = '__this_should_still_exist_';
        store2.persist(restored);
        await delay();

        expect(sessionStorage[TEST_KEY]).toMatch(/__this_should_still_exist_/);
    });
});

describe('ObjectStorage.persist', () => {
    it('should persist updated state', async () => {
        const data = { foo: 'bar' };
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('test', data);
        await delay();

        data.foo = 'baz';
        store.persist(data);
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        expect(store2.get('test')).toEqual({ foo: 'baz' });
    });
});

describe('ObjectStorage.delete', () => {
    it('should remove object from backing storage', async () => {
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('test', {});
        await delay();

        store.delete('test');
        expect(store.get('test')).toBeUndefined();
        expect(store.keys()).toEqual([]);
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        expect(store2.get('test')).toBeUndefined();
        expect(store2.keys()).toEqual([]);
    });

    it('should remove all unreachable objects from backing storage', async () => {
        const data1 = { a: '__this_should_still_exist_' };
        const data2 = { a: '__this_should_not_exist_' };
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('test1', { a: data1, b: data2 });
        store.set('test2', { a: data1 });
        await delay();

        store.delete('test1');
        await delay();
        expect(sessionStorage[TEST_KEY]).toMatchSnapshot();
    });

    it('should be no-op if key does not exist', async () => {
        const data = {};
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('data', data);
        await delay();

        store.delete('data');
        await delay();

        sessionStorage.clear();
        store.delete('data');
        await delay();
        expect(sessionStorage.length).toBe(0);
    });

    it('should mark line slot as empty', async () => {
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('data1', {});
        store.set('data2', {});
        await delay();
        expect(sessionStorage[TEST_KEY].split('\n').length).toBe(3);

        store.delete('data1');
        await delay();
        expect(sessionStorage[TEST_KEY].split('\n').length).toBe(3);

        store.set('data3', {});
        await delay();
        expect(sessionStorage[TEST_KEY].split('\n').length).toBe(3);
    });

    it('should not double free empty line', async () => {
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('data', [{}, {}]);
        await delay();

        const obj = {};
        store.set('obj', obj);
        store.delete('data');
        await delay();

        store.persist(obj);
        await delay();

        store.set('data', [
            { a: '__item_1__' },
            { a: '__item_2__' },
            { a: '__item_3__' },
            { a: '__item_4__' }
        ]);
        await delay();
        expect(sessionStorage[TEST_KEY]).toMatchSnapshot();
    });
});

describe('ObjectStorage.clear', () => {
    it('should clear all data', async () => {
        const store = createObjectStorage(sessionStorage, TEST_KEY);
        store.set('test', {});
        await delay();

        store.clear();
        expect(store.get('test')).toBeUndefined();
        expect(store.keys()).toEqual([]);
        await delay();

        const store2 = createObjectStorage(sessionStorage, TEST_KEY);
        expect(store2.keys()).toEqual([]);
    });
});
