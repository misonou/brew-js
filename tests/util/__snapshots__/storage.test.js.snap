// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectStorage.delete should not double free empty line 1`] = `
"{\\"obj\\":4,\\"data\\":1}
[\\"#2\\",\\"#3\\",\\"#5\\",\\"#6\\"]
{\\"a\\":\\"__item_1__\\"}
{\\"a\\":\\"__item_2__\\"}
{}
{\\"a\\":\\"__item_3__\\"}
{\\"a\\":\\"__item_4__\\"}"
`;

exports[`ObjectStorage.delete should remove all unreachable objects from backing storage 1`] = `
"{\\"test2\\":2}

{\\"a\\":\\"#3\\"}
{\\"a\\":\\"__this_should_still_exist_\\"}"
`;
